File: aaa/transpiler/code.py
class Code:
    def __init__(
        self, line: str | None = None, unindent: int = 0, indent: int = 0
    ) -> None:
        self.lines: list[str] = []
        self.indent_level = 0

        if line is not None:
            self.__add_line(line, bool(unindent), bool(indent))

    def __add_line(self, line: str, unindent: bool, indent: bool) -> None:
        if line.endswith("\n"):
            raise ValueError("Line should not end with newline.")

        if unindent:
            self.indent_level -= 1

        if self.indent_level < 0:
            raise ValueError("Indentation can't be negative.")

        indented_line = "    " * self.indent_level + line
        self.lines.append(indented_line)

        if indent:
            self.indent_level += 1

    def __add_code(self, code: "Code") -> None:
        for line in code.lines:
            self.__add_line(line, False, False)

    def add(self, added: "str | Code", unindent: int = 0, indent: int = 0) -> None:
        if isinstance(added, str):
            self.__add_line(added, bool(unindent), bool(indent))
        elif isinstance(added, Code):
            self.__add_code(added)

    def add_joined(self, join_line: str, joined: list["Code"]) -> None:
        for offset, line in enumerate(joined):
            self.add(line)
            if offset != len(joined) - 1:
                self.add(join_line)

    def get(self) -> str:
        return "\n".join(self.lines) + "\n"

File: aaa/src/transpiler/code.rs
const INDENTATION: &str = "    ";

pub struct Code {
    lines: Vec<String>,
    indent_level: usize,
}

impl Code {
    pub fn new() -> Self {
        Self {
            lines: vec![],
            indent_level: 0,
        }
    }

    pub fn from_string<T: Into<String>>(string: T) -> Self {
        let mut code = Self::new();
        code.add_line(string);
        code
    }

    pub fn add_line<T: Into<String>>(&mut self, value: T) {
        let string: String = value.into();

        if string.ends_with("\n") {
            panic!("add_line() string should not end with newline!")
        }

        if string.ends_with('}') || string.ends_with("},") {
            self.unindent();
        }

        let prefix = INDENTATION.repeat(self.indent_level);
        let line = format!("{}{}", prefix, string);

        self.lines.push(line);

        if string.ends_with('{') {
            self.indent();
        }
    }

    pub fn add_code(&mut self, code: Code) {
        let prefix = INDENTATION.repeat(self.indent_level);

        for line in &code.lines {
            self.lines.push(format!("{}{}", prefix, line));
        }
    }

    pub fn indent(&mut self) {
        self.indent_level += 1;
    }

    pub fn unindent(&mut self) {
        if self.indent_level == 0 {
            panic!("Cannot indent below level 0.");
        }

        self.indent_level -= 1;
    }

    pub fn get(&self) -> String {
        self.lines.join("\n") + "\n"
    }
}
